### ISSUE TITLE
Plane.PointAt(float u,float v,float w) Function

### ISSUE THREAD

**Issue by bradrothenberg** [nTopology/juce_icarus#3]
@Ahbee or @ebwork -- I am converting the exoskeleton code from c# to c++ -- & they have a few functions from Rhino that we do not have -- 

One of them is Plane.PointAt(float u, float v float w) -- which takes 3 floats, u,v,w. This should give a point on the plane at  plane.origin + u*plane.xaxis + v*plane.yaxis + z*plane.zaxis

In Juce_Icarus/gl, we have cinders Plane.h & .cpp -- do we have enough info w/ the planes normal & distance to add a PointAt function?

**Comment by bradrothenberg** [comment]
would this be correct: 

'''c++
template<typename T>
Vec3T cinder::Plane<T>::pointAt(T u, T v, T w)
{
	T xAxis = mNormal.x * mDistance * u;
	T yAxis = mNormal.y * mDistance * v;
	T zAxis = mNormal.z * mDistance * w;

	return Vec3T(xAxis, yAxis, zAxis);
}
'''

**Comment by ebwork** [comment]
I am not at a computer right now (waiting for a routine test at the doctor)
but I will take a stab at this.

First the code you have won't work - easiest way to see is, suppose
mDistance is zeto, then all inputs give 0,0,0

I assume they are computing origin_plane + u * xaxis_plane + v *
yaxis_plane + w * zaxis_plane .  Note that the xaxis_plane etc are unit
vectors.  Does the cinder plane class keep these internally?

If you want to send the code for plane.pointAt and for the cinder plane
class I will look at it.  Will be using phone only for the next hour or two.

Erik

**Comment by bradrothenberg** [comment]
good call -- yeah, no rush, I can wait till you're back at the computer.

Cinders plane class only has two member variables, mNormal, & mDistance. We can add the unit vectors for x, y, & z if needed though...

**Comment by bradrothenberg** [comment]
I also added these functions to the plane class:

```c++
template<typename T>
Vec3T cinder::Plane<T>::getZAxis() const
{
	glm::vec3 zAxis;

	return zAxis;
}

template<typename T>
Vec3T cinder::Plane<T>::getYAxis() const
{
	glm::vec3 yAxis;

	return yAxis;
}

template<typename T>
Vec3T cinder::Plane<T>::getXAxis() const
{
	glm::vec3 xAxis;

	return xAxis;
}

```

But they need to be implemented

**Comment by Ahbee** [comment]
> One of them is Plane.PointAt(float u, float v float w) -- which takes 3 floats, u,v,w. This should give a point on the plane at plane.origin + uplane.xaxis + vplane.yaxis + z*plane.zaxis

Sorry I dont understand stand what you mean? 
a plane defined by a point and a normal. It does not have an origin, xAxis,yAxis,zAxis. 

I would have guessed that a function like Plane.pointAt  reorients the plane or something like that.

Can I see a link to the c sharp function?

**Comment by Ahbee** [comment]
Unless it is not a plane and a rectangle where the center is the origin and the normal is the zAxis?

**Comment by ebwork** [comment]
Defining the plane using the normal and distance is very concise, and is
good for doing a "which-side-of" test.  But is does not provide a unique
coordinate system on the plane, i.e., the xaxis, yaxis, and zaxis are not
uniquely defined.

Can you point me to the exoskeleton code?

On Mon, Jun 8, 2015 at 4:27 PM, Bradley Rothenberg <notifications@github.com

**Comment by Ahbee** [comment]
> Defining the plane using the normal and distance is very concise, and is good for doing a "which-side-of" test. But is does not provide a unique coordinate system on the plane, i.e., the xaxis, yaxis, and zaxis are not uniquely defined.

there is no coordinate system for a plane? what would it be? If you take the normal to be the z axis, then 
which do you take for the x and y axis?  Unless its a rectangle it does not make sense.

**Comment by bradrothenberg** [comment]
The Rhino PointAt function is here : 
http://4.rhino3d.com/5/rhinocommon/html/M_Rhino_Geometry_Plane_PointAt_1.htm

Good point -- I think there needs to be an additional piece of information to get the y axis, so iff you have Z & Y, you can get X -- 

the exoskeleton code is here:

https://github.com/davestasiuk/Exoskeleton2

currently, I'm gonig through hull:

    public static void Hull(List<Point3d> Pts, ref Mesh Msh)
        {

            List<int> Indices = new List<int>();
            Line HullStart = new Line(Pts[0], Pts[1]);
            Indices.Add(0); Indices.Add(1);
            
            bool PlaneSet = false;
            Plane PlaneStart = Plane.Unset;

            for (int P = 2; P < Pts.Count; P++)
            {
                if (!PlaneSet)
                {
                    if (HullStart.ClosestPoint(Pts[P],false).DistanceTo(Pts[P]) > RhinoDoc.ActiveDoc.ModelAbsoluteTolerance)
                    {
                        PlaneSet = true;
                        Indices.Add(P);
                        PlaneStart = new Plane(Pts[0], Pts[1], Pts[P]);
                    }
                }
                else
                {
                    if (PlaneStart.DistanceTo(Pts[P]) > RhinoDoc.ActiveDoc.ModelAbsoluteTolerance)
                    {
                        Indices.Add(P);
                        break;
                    }
                }
            }

            Indices.Reverse();

            foreach (int I in Indices)
            {
                Msh.Vertices.Add(Pts[I]);
                Pts.RemoveAt(I);
            }

            Msh.Faces.AddFace(new MeshFace(0, 1, 2));
            Msh.Faces.AddFace(new MeshFace(0, 2, 3));
            Msh.Faces.AddFace(new MeshFace(0, 3, 1));
            Msh.Faces.AddFace(new MeshFace(1, 2, 3));

            do
            {
                NormaliseMesh(ref Msh);
                GrowHull(ref Msh, Pts[0]);
                Pts.RemoveAt(0);
            } while (Pts.Count > 0);

            Msh.Vertices.CullUnused();

        }

**Comment by Ahbee** [comment]
yeah cause the rhino plane structure is different it takes two vectors and an origin 
http://4.rhino3d.com/5/rhinocommon/html/T_Rhino_Geometry_Plane.htm

I would just create another data structure called NTPlane to mimic rhinos, don't mess with cinders. Cause the two are inherently different data structures.

**Comment by bradrothenberg** [comment]
Ah yeah I see that -- ok that makes sense

**Comment by bradrothenberg** [comment]
I think we can do what exoskeleton is doing using our tube class, & the parallel transport frame -- the exoskeleton is relying on this function curve.extractFrameAt(float t) -- which we don't have. But we do have the tube class which already builds these frames. so as long as we can store the start & end frames in a list, we should be able to convex hul lthose pts, & be good...
